tutorials/first-steps.md

```diff
@@ -0,0 +1,397 @@
+> - 4️⃣ `java.lang.Boolean`
+> - 5️⃣ `java.lang.String`
+> - 7️⃣ `java.lang.Integer`
+> - 8️⃣ `java.lang.Double`
+> - 9️⃣ `java.lang.Double`
+> - 1️⃣0️⃣ you'll get `46.2`
+> - 1️⃣1️⃣ Golo identifiers can be non-ascii characters (e.g., Japanese, Emoji, Arabic, etc)
+> - 1️⃣2️⃣ you'll get `oh oh, it's bad`
+
+✋ **To run it**, type this command: `golo golo --files 02-variables-and-co.golo`
+
+## Functions
+
+Functions are first-class citizen in Golo. And there are various way to define these functions.
+
+Create a new `03-functions.golo` file:
+
+> **03-functions.golo**
+```golo
+module hello.world
+
+----
+This function takes no argument
+and return a string 0️⃣
+----
+function hello = { # 1️⃣
+  return "Hello world!"
+}
+
+----
+This function takes one argument
+and return a string
+----
+function hello_who = |name| { # 2️⃣
+  return "Hello " + name
+}
+
+----
+This function takes two arguments
+and return a string
+----
+function hello_people = |name_1, name_2| -> "Hello " + name_1 + " and " + name_2 # 3️⃣
+
+----
+This is my main function
+----
+function main = |args| { # 4️⃣
+  println(hello())
+  println(hello_who("Bob"))
+  println(hello_people("Bob", "Bill"))
+}
+
+```
+
+> - 0️⃣ you can use multi-lines comments at the header of the functions
+> - 1️⃣ this is a parameter-less function
+> - 2️⃣ you need to use the pipe notation to pass parameter(s)
+> - 3️⃣ you can use a compact form (`|arg| -> 42`) to define a regular function (you don't need to use the return keyword)
+> - 4️⃣ you'll get:
+>    ```
+>    Hello world!
+>    Hello Bob
+>    Hello Bob and Bill
+>    ```
+
+✋ **To run it**, type this command: `golo golo --files 03-functions.golo`
+
+ℹ️ more about functions: [https://golo-lang.org/documentation/next/index.html#_functions](https://golo-lang.org/documentation/next/index.html#_functions).
+
+## Closures
+
+**From the Golo documentation**: *"Golo supports closures, which means that functions can be treated as first-class citizen. Defining a closure is straightforward as it derives from the way a function can be defined."*
+
+Create a new `04-closures.golo` file:
+
+> **04-closures.golo**
+```golo
+module hello.world
+
+function main = |args| {
+
+  let say_hello = -> println("Hello") # I know it's not pure 😝
+
+  say_hello() # 1️⃣
+
+  let say_hello_to_somebody = |name| {
+    println("Hello " + name)
+  }
+
+  say_hello_to_somebody("Bob") # 2️⃣
+
+  let sum = |a, b| -> a + b 3️⃣
+
+  println(
+    sum(10, sum(30, 2))
+  ) # 4️⃣ it will print 42
+
+}
+```
+
+> - 1️⃣ it will print `Hello`
+> - 2️⃣ it will print `Hello Bob`
+> - 3️⃣ Golo supports a compact form of closures for the cases where their body consists of a single expression
+> - 4️⃣ it will print `42`
+
```

Comment:
`--pedantic` these are not closures, since no free variable is closed from the context of the lambda definition.
Besides the semantic discussion, maybe a example of a true closure could be useful, e.g.
```golo
  let adder = |first| -> |second| -> first + second

  let add_two = adder(2)
  let add_four = adder(4)

  println(add_two(40))
  println(add_four(38))
```
or maybe clearer written
```golo
let adder = |first| {
  return |second| -> first + second
}
```

Reply:
it comes from the Golo documentation

Reply:
Yes, the "closure" term is a bit overloaded :wink: